flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/RetryTestExecutionExtension.java

```diff
@@ -0,0 +1,121 @@
+        Map<String, Boolean> shouldRetry =
+                (Map<String, Boolean>) context.getStore(RETRY_NAMESPACE).get(RETRY_KEY);
+        String method = getTestMethodKey(context);
+        if (!shouldRetry.get(method)) {
+            return ConditionEvaluationResult.disabled(method + "have already passed or failed.");
+        }
+        return ConditionEvaluationResult.enabled(
+                String.format("Test %s: [%d/%d]", method, retryIndex, totalTimes));
+    }
+
+    @Override
+    public void handleTestExecutionException(ExtensionContext context, Throwable throwable)
+            throws Throwable {
+        String method = getTestMethodKey(context);
+        if (repeatableException != null) {
+            // RetryOnException
+            if (repeatableException.isAssignableFrom(throwable.getClass())) {
+                // 1. continue retrying when get some repeatable exceptions
+                String retryMsg =
+                        String.format(
+                                "Retry test %s[%d/%d] failed with repeatable exception, continue retrying.",
+                                method, retryIndex, totalTimes);
+                LOG.error(retryMsg, throwable);
+                throw new TestAbortedException(retryMsg);
+            } else {
+                // 2. stop retrying when get an unrepeatable exception
+                Map<String, Boolean> shouldRetry =
+                        (Map<String, Boolean>) context.getStore(RETRY_NAMESPACE).get(RETRY_KEY);
+                shouldRetry.put(method, false);
+                LOG.error(
+                        String.format(
+                                "Retry test %s[%d/%d] failed with unrepeatable exception, stop retrying.",
+                                method, retryIndex, totalTimes),
+                        throwable);
+                throw throwable;
+            }
+        } else {
+            // RetryOnFailure
+            // 1. Failed when reach the total retry times
+            if (retryIndex == totalTimes) {
+                LOG.error("Test Failed at the last retry.", throwable);
+                throw throwable;
+            }
+
+            // 2. continue retrying
+            String retryMsg =
+                    String.format(
+                            "Retry test %s[%d/%d] failed, continue retrying.",
+                            method, retryIndex, totalTimes);
+            LOG.error(retryMsg, throwable);
+            throw new TestAbortedException(retryMsg);
+        }
+    }
+
+    @Override
+    public void afterEach(ExtensionContext context) throws Exception {
+        Throwable exception = context.getExecutionException().orElse(null);
+        if (exception == null) {
+            Map<String, Boolean> shouldRetry =
+                    (Map<String, Boolean>) context.getStore(RETRY_NAMESPACE).get(RETRY_KEY);
+            String method = getTestMethodKey(context);
+            shouldRetry.put(method, false);
+            LOG.info(
```

Comment:
```suggestion
            LOG.trace(
```
This will just be noise.